File: CS/ComplexDialogSample.Module/Controllers/MyController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Editors;
using ComplexDialogSample.Module.BusinessObjects;

namespace ComplexDialogSample.Module.Controllers {

    [DevExpress.ExpressApp.DC.DomainComponent]
    public class OrderTemplate {
        public OrderTemplate(Session s) {
            _Services = new XPCollection<Service>(s);
        }
        public DateTime DueDate { get; set; }
        public Team Team { get; set; }
        private XPCollection<Service> _Services;
        public XPCollection<Service> Services { get { return _Services; } }
    }

    public class MyController : ViewController {
        public MyController() {
            TargetObjectType = typeof(Office);
            TargetViewType = ViewType.ListView;
            PopupWindowShowAction action = new PopupWindowShowAction(this, "AssignJobs", PredefinedCategory.RecordEdit);
            action.SelectionDependencyType = SelectionDependencyType.RequireMultipleObjects;
            action.CustomizePopupWindowParams += new CustomizePopupWindowParamsEventHandler(action_CustomizePopupWindowParams);
            action.Execute += new PopupWindowShowActionExecuteEventHandler(action_Execute);
        }
        void action_CustomizePopupWindowParams(object sender, CustomizePopupWindowParamsEventArgs e) {
            IObjectSpace os = Application.CreateObjectSpace();
            e.Context = TemplateContext.PopupWindow;
            e.View = Application.CreateDetailView(os, new OrderTemplate(((DevExpress.ExpressApp.Xpo.XPObjectSpace)os).Session));
            ((DetailView)e.View).ViewEditMode = ViewEditMode.Edit;
        }
        void action_Execute(object sender, PopupWindowShowActionExecuteEventArgs e) {
            OrderTemplate parameters = e.PopupWindow.View.CurrentObject as OrderTemplate;
            ListPropertyEditor listPropertyEditor = ((DetailView)e.PopupWindow.View).FindItem("Services") as ListPropertyEditor;
            IObjectSpace os = Application.CreateObjectSpace();
            foreach (Office b in e.SelectedObjects) {
                Team team = os.GetObject<Team>(parameters.Team);
                foreach (Service service in listPropertyEditor.ListView.SelectedObjects) {
                    Order order = os.CreateObject<Order>();
                    order.DueDate = parameters.DueDate;
                    order.Team = team;
                    order.Office = os.GetObject<Office>(b);
                    order.Service = os.GetObject<Service>(service);
                    order.Save();
                }
            }
            os.CommitChanges();
        }
    }
}

File: VB/ComplexDialogSample.Module/Controllers/MyController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Editors
Imports ComplexDialogSample.Module.BusinessObjects

Namespace ComplexDialogSample.Module.Controllers

    <DevExpress.ExpressApp.DC.DomainComponent> _
    Public Class OrderTemplate
        Public Sub New(ByVal s As Session)
            _Services = New XPCollection(Of Service)(s)
        End Sub
        Public Property DueDate() As Date
        Public Property Team() As Team
        Private _Services As XPCollection(Of Service)
        Public ReadOnly Property Services() As XPCollection(Of Service)
            Get
                Return _Services
            End Get
        End Property
    End Class

    Public Class MyController
        Inherits ViewController

        Public Sub New()
            TargetObjectType = GetType(Office)
            TargetViewType = ViewType.ListView
            Dim action As New PopupWindowShowAction(Me, "AssignJobs", PredefinedCategory.RecordEdit)
            action.SelectionDependencyType = SelectionDependencyType.RequireMultipleObjects
            AddHandler action.CustomizePopupWindowParams, AddressOf action_CustomizePopupWindowParams
            AddHandler action.Execute, AddressOf action_Execute
        End Sub
        Private Sub action_CustomizePopupWindowParams(ByVal sender As Object, ByVal e As CustomizePopupWindowParamsEventArgs)
            Dim os As IObjectSpace = Application.CreateObjectSpace(GetType(Service))
            e.Context = TemplateContext.PopupWindow
            e.View = Application.CreateDetailView(os, New OrderTemplate(DirectCast(os, DevExpress.ExpressApp.Xpo.XPObjectSpace).Session))
            CType(e.View, DetailView).ViewEditMode = ViewEditMode.Edit
        End Sub
        Private Sub action_Execute(ByVal sender As Object, ByVal e As PopupWindowShowActionExecuteEventArgs)
            Dim parameters As OrderTemplate = TryCast(e.PopupWindow.View.CurrentObject, OrderTemplate)
            Dim listPropertyEditor As ListPropertyEditor = TryCast(CType(e.PopupWindow.View, DetailView).FindItem("Services"), ListPropertyEditor)
            Dim os As IObjectSpace = Application.CreateObjectSpace(GetType(Team))
            For Each b As Office In e.SelectedObjects
                Dim team As Team = os.GetObject(Of Team)(parameters.Team)
                For Each service As Service In listPropertyEditor.ListView.SelectedObjects
                    Dim order As Order = os.CreateObject(Of Order)()
                    order.DueDate = parameters.DueDate
                    order.Team = team
                    order.Office = os.GetObject(Of Office)(b)
                    order.Service = os.GetObject(Of Service)(service)
                    order.Save()
                Next service
            Next b
            os.CommitChanges()
        End Sub
    End Class
End Namespace
